### PR TITLE
Update controller mapping for Behringer DDM4000 mixer

### PR DESCRIPTION
**Pull request by git-developer**
This PR contains a few fixes and updates for the controller mapping of the Behringer DDM4000 mixer:
* `show_xfader` is no longer in group `[Master]` but in group `[Skin]`
* The `ShiftButton` supports default button behavior (by deriving `Button`s `input` function)
* The `DirectionEncoder` supports a relative mode (as known for [Pots](https://github.com/mixxxdj/mixxx/wiki/Components-JS#pot))
* New component `RangeAwarePot` that combines the behavior of a `RangeAwareEncoder` with soft-takeover
* Controller feedback for `BackLoopButton`
* Allow `sendShifted` for effect components
* Minor fixes and updates to the documentation

The changes are fully backwards-compatible (except the `show_xfader` fix, of course). They have been tested extensively with a custom mapping that maps an effect unit and deck effects (loop, reverse) to the mic, crossfader and sampler sections of the mixer. I don't plan to merge this custom mapping here (because the current mapping resembles the mixer functions best), but I plan to share it on the forums when this PR is accepted. I can provide the mapping if desired.

There's intentionally no PR for the manual because the changes do not affect the documentation.

### CODE REVIEW

**Comment by uklotzde** [comment]
We also need a PR for updating the manual accordingly. Both should be merged together.

**Comment by git-developer** [comment]
I'd love to add a PR for the manual, but there's no change in the documentation involved. All implementation changes are internally; just a few bug fixes and additional features that may be used by custom mappings.

File: res/controllers/Behringer-Extension-scripts.js [hunk 1]
```diff
@@ -105,12 +108,13 @@
...
     };
     ShiftButton.prototype = deriveFrom(components.Button, {
-        input: function(_channel, _control, value, _status, _group) {
+        inSetValue: function(value) {
             if (value) {
                 this.target.shift();
             } else {
                 this.target.unshift();
             }
+            components.Button.prototype.inSetValue.call(this, value);
```
Comment (Swiftb0y) [inline res/controllers/Behringer-Extension-scripts.js:117]: can you explain what the purpose of this change is?

File: res/controllers/Behringer-Extension-scripts.js [hunk 2]
```diff
@@ -289,6 +293,14 @@
      * Turning the encoder to the right means "forwards" and returns 1;
      * turning it to the left means "backwards" and returns -1.
      *
+     * This component support an optional relative mode as an alternative to
```
Comment (Swiftb0y) [inline res/controllers/Behringer-Extension-scripts.js:296]:
```suggestion
     * This component supports an optional relative mode as an alternative to
```

File: res/controllers/Behringer-Extension-scripts.js [hunk 3]
```diff
@@ -467,13 +502,13 @@
...
-        options.outKey = options.outKey || "loop_enabled";
+        options.key = options.key || "loop_enabled";
         components.Button.call(this, options);
     };
     BackLoopButton.prototype = deriveFrom(components.Button, {
-        input: function(_channel, _control, value, _status, group) {
-            var engine = global.engine;
+        inSetValue: function(value) {
             var script = global.script;
+            var group = this.group;
```
Comment (Swiftb0y) [inline res/controllers/Behringer-Extension-scripts.js:511]: why is this line required?

File: res/controllers/Behringer-Extension-scripts.js [hunk 4]
```diff
@@ -1307,6 +1348,14 @@
                 new components.EffectUnit(effectUnitDefinition.unitNumbers, true),
                 componentStorage,
                 ["onFocusChange", "shift", "unshift"]);
+            var shiftType = effectUnitDefinition.sendShiftedFor;
+            if (typeof shiftType === "function") {
+                effectUnit.forEachComponent(function(component) {
+                    if (component instanceof shiftType) {
+                        component.sendShifted = true;
```
Comment (Swiftb0y) [inline res/controllers/Behringer-Extension-scripts.js:1355]: whats the `shiftType === "function"` for?

**Review by Swiftb0y — CHANGES_REQUESTED** [review]

File: res/controllers/Behringer-Extension-scripts.js (hunk 1 above)
Comment (git-developer) [reply res/controllers/Behringer-Extension-scripts.js:117]: I mapped the Shift button to a physical button that has a LED, and noticed that the LED does not work. This was caused by Mixxx not sending anything when the `outKey` control changed, because the input function of `Component` was overridden.

The current change does not override `input` so the original implementation survives, and delegates to the `Button`'s `inSetValue` function so that Shift behaves just as any other button does.

File: res/controllers/Behringer-Extension-scripts.js (hunk 3 above)
Comment (git-developer) [reply res/controllers/Behringer-Extension-scripts.js:511]: This is required because `group` is no longer an argument if the surrounding function. `group` is used 7 times within the function, so it either has to be declared like this or I'd have to change the 7 occurrences to `this.group`.

File: res/controllers/Behringer-Extension-scripts.js (hunk 4 above)
Comment (git-developer) [reply res/controllers/Behringer-Extension-scripts.js:1355]: The property `sendShiftedFor` is a filter tor the type of a JS component , e.g. `c.Button`, `c.Pot` or `c.Component`. If the user provides a string, number, boolean or object instead, the following `instanceof` statement will cause a runtime error.

File: res/controllers/Behringer-Extension-scripts.js (hunk 1 above)
Comment (Swiftb0y) [reply res/controllers/Behringer-Extension-scripts.js:117]: Sorry I'm still confused (likely because I have 0 idea how the accompanying library you wrote works). 
The shift button does not seem to be coupled to any CO, so the `outKey` is irrelevant.

you can still overwrite your input implemenation, you just need to send the midi output yourself using `this.send(value)` for example.

File: res/controllers/Behringer-Extension-scripts.js (hunk 3 above)
Comment (Swiftb0y) [reply res/controllers/Behringer-Extension-scripts.js:511]: Thanks, makes sense. I guess thats valid.

File: res/controllers/Behringer-Extension-scripts.js (hunk 4 above)
Comment (Swiftb0y) [reply res/controllers/Behringer-Extension-scripts.js:1355]: Mhmm makes sense, but its not self explaining. Could you add a comment that explains this?

File: res/controllers/Behringer-Extension-scripts.js (hunk 1 above)
Comment (git-developer) [reply res/controllers/Behringer-Extension-scripts.js:117]: OK, let me explain it a little bit more detailed; you don't need to understand the whole library for that.

The shift button is first of all a just a normal `Button` that is connected to a user-defined control and additionally calls `shift()` and `unshift()` on some other ("target") component. The user may configure the connected Mixxx control as well as other options (like the button's `type`, for example).

In the [BCR2000 mapping](https://github.com/mixxxdj/mixxx/blob/989f7c8474c2b22dcc31bc00b089f3f09ca3c2b0/res/controllers/Behringer-BCR2000-scripts.js#L25), the connected Mixxx control is configured as `group: [Controls]` & `key: touch_shift`, and the target is the top level `ComponentContainer` that contains all components of the mapping. When the user presses the shift button, `shift()` is called on the top level container and then delegated to all child components. This is effectively a shift for all components of the whole mapping. When the user's skin contains a control for `[Controls],touch_shift`, the shift state is also visualized in the UI.

I didn't use a default for `group` and `key` in the `ShiftButton`'s constructor because it may as well be used for just a group of components or just for a single component. `[Controls],touch_shift` would be misleading in that case.

It's true that I could override `input()` and call `this.send(value)` instead of overriding `inSetValue()`. But when I do so, the `ShiftButton` loses the feature of button types (`push`, `toggle`, `powerWindow`), because this is implemented in `Button.input()`.

File: res/controllers/Behringer-Extension-scripts.js (hunk 4 above)
Comment (git-developer) [reply res/controllers/Behringer-Extension-scripts.js]: Of course, done.

**Comment by uklotzde** [comment]
@Swiftb0y Merge?

File: res/controllers/Behringer-Extension-scripts.js (hunk 1 above)
Comment (Swiftb0y) [reply res/controllers/Behringer-Extension-scripts.js:117]: Ah now I see, thanks for your explanation. I'm sorry I forced you to write this detailed explanation.

**Review by Swiftb0y — APPROVED** [review]
